Sources/StreamChat/StateLayer/UserSearchState.swift

```diff
@@ -27,23 +23,40 @@ extension UserSearchState {
     @MainActor func value<Value>(forKeyPath keyPath: KeyPath<UserSearchState, Value>) -> Value {
         self[keyPath: keyPath]
     }
+    
+    @MainActor private func setValue<Value>(_ value: Value, for keyPath: ReferenceWritableKeyPath<UserSearchState, Value>) {
+        self[keyPath: keyPath] = value
+    }
         
-    @MainActor func setActiveTask(_ task: Task<[ChatUser], Error>, query: UserListQuery) {
-        if let activeTask {
-            activeTask.cancel()
-        }
-        activeTask = task
+    @MainActor func handleStartingFetchingQuery(_ query: UserListQuery) {
         self.query = query
     }
-    
-    @MainActor func setUsers(_ newUsers: [ChatUser], for query: UserListQuery, pagination: Pagination) {
-        // When the query changes we set the pagination to 0, otherwise we are loading more results for the same query
-        if pagination.offset == 0 {
-            users = StreamCollection(newUsers)
+
+    func handleFinishedFetchingQuery(_ completedQuery: UserListQuery, users incomingUsers: [ChatUser]) async {
+        if let query = await value(forKeyPath: \.query), query.hasFilterOrSortingChanged(completedQuery) {
+            // Discard since filter or sorting has changed
+            return
+        }
+        let result: StreamCollection<ChatUser>
+        if completedQuery.pagination?.offset == 0 {
+            // Reset to the first page
+            result = StreamCollection(incomingUsers)
         } else {
-            var result = Array(users[..<pagination.offset])
-            result.append(contentsOf: newUsers)
-            users = StreamCollection(result)
+            // Filter and sorting are the same but incoming users might contain duplicates (depends how pagination is used)
+            let incomingIds = Set(incomingUsers.map(\.id))
+            let incomingRemovedUsers = users.filter { !incomingIds.contains($0.id) }
+            let sortValues = completedQuery.sort.map(\.sortValue)
+            result = StreamCollection((incomingRemovedUsers + incomingUsers).sort(using: sortValues))
```

Comment:
How can they contain duplicates? 🤔

Reply:
If you call `loadNextUsers` multiple times.